parlai/agents/starspace/starspace.py

```diff
@@ -0,0 +1,459 @@
+                           'training (may reduce accuracy). This fixes all '
+                           'input and output to have a maximum length.')
+        agent.add_argument('-k', '--neg-samples', type=int, default=10,
+                           help='number k of negative samples per example')
+        agent.add_argument('--parrot-neg', type=int, default=0,
+                           help='include query as a negative')
+        agent.add_argument('--tfidf', type='bool', default=False,
+                           help='Use frequency based normalization for embeddings.')
+        agent.add_argument('-cs', '--cache-size', type=int, default=1000,
+                           help='size of negative sample cache to draw from')
+        agent.add_argument('-hist', '--history-length', default=10000, type=int,
+                           help='Number of past tokens to remember. ')
+        agent.add_argument('-histr', '--history-replies',
+                           default='label', type=str,
+                           choices=['none', 'model', 'label'],
+                           help='Keep replies in the history, or not.')
+        agent.add_argument('-fixedCands', '--fixed-candidates-file',
+                           default=None, type=str,
+                           help='File of cands to use for prediction')
+
+    def __init__(self, opt, shared=None):
+        """Set up model if shared params not set, otherwise no work to do."""
+        super().__init__(opt, shared)
+        opt = self.opt
+        self.reset_metrics()
+        self.id = 'Starspace'
+        self.NULL_IDX = 0
+        self.cands = torch.LongTensor(1, 1, 1)
+        self.ys_cache = []
+        self.ys_cache_sz = opt['cache_size']
+        self.truncate = opt['truncate'] if opt['truncate'] > 0 else None
+        self.history = {}
+        if shared:
+            self.threadindex = shared['threadindex']
+            print("[ creating Starspace thread " + str(self.threadindex)  + " ]")
+            # set up shared properties
+            self.dict = shared['dict']
+            self.model = shared['model'] #Starspace(opt, len(self.dict))
+        else:
+            print("[ creating StarspaceAgent ]")
+            # this is not a shared instance of this class, so do full init
+            if opt['dict_file'] is None and opt.get('model_file'):
+                # set default dict-file if not set
+                opt['dict_file'] = opt['model_file'] + '.dict'
+            # load dictionary and basic tokens & vectors
+            self.dict = DictionaryAgent(opt)
+
+            self.model = Starspace(opt, len(self.dict), self.dict)
+            if opt.get('model_file') and os.path.isfile(opt['model_file']):
+                self.load(opt['model_file'])
+            self.model.share_memory()
+
+        # set up modules
+        self.criterion = torch.nn.CosineEmbeddingLoss(margin=opt['margin'], size_average=False)
```

Comment:
just curious--why `size_average=False`? I don't have any intuition either way

Reply:
i think it's just a scaling factor based on batch size? actually i only  batch size 1 right now anyway..